### PR TITLE
Raul - New Timer System

### PR DESCRIPTION
**Pull request by talDoFlemis**
# Description

PLEASE, DO NOT SUGGEST NEW FEATURES AND CHANGES, JUST TEST AND VERIFY IF IT'S WORKING. SUGGEST ONLY FIXES!

Please tag @rauleffting , when you find something, this user is not currently on the project.

This PR implements the New Timer system that can now synchronize between multiples tabs, detect whenever the user is inactive and adds a new way to track the time with a Countdown mode, while improving speed of the connection and reducing the load to MongoDB.

## Related PRS (if any):

This PR is related to this PR in [backend](https://github.com/OneCommunityGlobal/HGNRest/pull/271)

## Mainly changes explained:

* Add react-use-websocket and progressbar dependencies
* Update timer-service endpoint 
* Update Header to use new timer 
* Add NewTimer, Countdown, Stopwatch, SwitchTimer and TimerStatus Components

## How to test:

Make sure you have the most updated branch,if you have some conflict no worries, don't need to solve ! Just do "git checkout -f development && git branch -D  said-feat-realtime-clock && git fetch --prune" (This command will delete the current branch of you machine and pull the branch again fully updated)

* Checkout to this branch in both frontend and backend
* Run `npm install` on the frontend to install new dependencies and `npm start`
* Run `npm install` on the backend to install new dependencies, `npm build` to build the server and `npm start` to start the server
* Log as any user
* Verify new timer running on the left of Navbar
* Run the stopwatch or countdown
* Open a new tab and sees that the timer is running with the same amount of time elapsed
* It works even on anonymous tabs, but without sync.

## Screenshots or videos of changes:
[Timer Showdown Video](https://www.dropbox.com/s/ybwo1z07flkk0zt/lastPreview.mp4?dl=0)
[User Stop Condition Video](https://www.dropbox.com/s/fccti51ftj5x01y/userStopsCondition.mp4?dl=0)
image
image

## Note:
This is a draft PR and any optimizations coming from the Seniors are welcome, this is one of the core feature of the HGN.

### CODE REVIEW

**Comment by Filipe-Oliv** [comment]
Hi @talDoFlemis 

I identified that it is not possible to switch between stopwatch and countdown;

As for the rest, the entire interface looks great and the countdown is working properly.

[week 03 - 02-11-23 (#653).webm](https://user-images.githubusercontent.com/114119668/218268565-11e35eb9-c303-480d-8d63-971658792026.webm)

**Comment by talDoFlemis** [comment]
The team requested to disable the stopwatch feature from this new timer, but it source code still on this PR if someone decides to enable it again

> Hi @talDoFlemis
> 
> I identified that it is not possible to switch between stopwatch and countdown;
> 
> As for the rest, the entire interface looks great and the countdown is working properly.
> 
>  week.03.-.02-11-23.653.webm

**Comment by NyahJ** [comment]
Hi @talDoFlemis,

I have followed your suggestions to test the functions and features of the count down modal and I recorded two videos for the testing. 
The first one is to start the count down, everything worked as intended.
https://user-images.githubusercontent.com/77643813/222009704-f3d91d8f-ed97-4244-b4b6-3cbfa0ad75bf.mp4

The second one is the ending of the count down, a small issue arises: Since the final second ends, the "Add Tangible Time Entry" appears, asking users to log their time. Since the time to be logged is fixed by the count down, why do we have these 3 buttons here in the time entry page? I believe it will add the time for count down for the next round, which does not make any sense here?
截屏2023-02-28 下午4 11 14

 截屏2023-02-28 下午4 07 11
https://user-images.githubusercontent.com/77643813/222010890-5edd308e-b7e4-4925-a99c-0f2abc64e836.mp4

**Comment by one-community** [comment]
This is a Draft/unfinished PR, so no need to review it. I've got Harlley and Vanel taking over finishing this now though, so we'll be reaching out to the whole team to test it once we've got a test version up and running.

**Comment by harlleybastos** [comment]
Just to notice, now I finish the last edits, it's okay to start to fix. I will for all your thoughts about that feature.

**Review by kangjay88 — COMMENTED** [review]
Hi @talDoFlemis, I have tested this PR on my local server using BE 271, and npm install. I am having trouble looking at the timer functionality as seen in my video. May we get a demonstration on how the timer should function now with your edits (I think the dropbox videos have been deleted)? I see the message: 'Waiting for messages from server' which may be from the TimerStatus component (0:30 sec mark). Not sure if anyone else is running into this error!


https://user-images.githubusercontent.com/108297706/226137353-8961cbcc-a018-4bd5-94bf-4ec37527b497.mp4



PR653 Code

**Comment by harlleybastos** [comment]
@NataliaCavicchioli and @kangjay88, I just tested the branch again here and it's everything working as it should. Please do "git fetch --prune". On the video I explain what this command do.

https://user-images.githubusercontent.com/37200778/226678721-eb973f3f-f5b2-4b8f-8361-330e721a2ecd.mp4

**Comment by NataliaCavicchioli** [comment]
> @NataliaCavicchioli and @kangjay88, I just tested the branch again here and it's everything working as it should. Please do "git fetch --prune". On the video I explain what this command do. Do that on both, frontend and backend branch.

Hey @harlleybastos! I did what you suggested, on FE and BE, and I'm sorry to let you know that the bug remains. 

I am on branch said-feat-realtime-clock on both front-end and back-end, and the error messages I am getting on console are related to variable values. 

From what I could understand, the problem is that these variables: timer, resetTimer, goal, handleAddGoal, handleStop, are getting to TimeEntryForm as undefined. I couldn't understand why, but the application is breaking because of it.

It would be great if someone else could test it to see if they have the same errors I'm having.

image

**Comment by AriaYu927** [comment]
Hi @harlleybastos , I've tested the timer and it overall works excellent! But I found a small issue: the default goal of the day is 0:0 if it's the first time the user use the countdown. Then, the countdown can still start if the goal is 0:0. Finally, when I stop the timer and try to add tangible time entry, I see the value of the final duration is minus. Here's the video I recorded:

https://user-images.githubusercontent.com/94303659/227375709-c1cbd643-fdc2-4cb3-96de-364da56444b7.mp4

Besides, I didn't get the error that @NataliaCavicchioli and you discussed.
And thank you for your effort on this feature, I really like the design!

**Comment by harlleybastos** [comment]
> Hi @harlleybastos , I've tested the timer and it overall works excellent! But I found a small issue: the default goal of the day is 0:0 if it's the first time the user use the countdown. Then, the countdown can still start if the goal is 0:0. Finally, when I stop the timer and try to add tangible time entry, I see the value of the final duration is minus. Here's the video I recorded:
> 
> Besides, I didn't get the error that @NataliaCavicchioli and you discussed. And thank you for your effort on this feature, I really like the design!

Thanks for let me know, I will add here the list of bugs that I need to look at, and about the problem of @NataliaCavicchioli I'm helping her to debug.

**Review by PrabhjotSembhi — COMMENTED** [review]
Here are my review and comments:

1.  when the computer goes to sleep timer gets paused. but if I keep it unattended for a while it doesn't pause. I tried for 15 unattended and 30 mins unattended too But it doesn't pause. 

pause


2. When I am trying to log my time I am not saying any project/task in the dropdown.
Screenshot 2023-03-25 173858


3. It's not alerting me if I am working and my timer is paused. I have paused it manually 


4. The timer we are using now shows the time that has been logged/spent and this shows the countdown timer. If we can add logged/spent hours to that will be better. (this is my personal opinion) We can either add that in red highlight space or on top of the clock.

5. I got an alert with a sound when the countdown is finished. 


6. I think we can reduce the font size of countdown and add little padding between numbers and text ( minutes and Seconds )

**Review by kangjay88 — COMMENTED** [review]
Hi @harlleybastos , the timer is working for me now after committing the new changes. Based on your message on Slack, I assume Raul's suggestions are already in progress.
So I will try to follow Jae's criteria for the timer:
1. Beyond Raul's comment about timer pausing on projects page, The timer pauses when accessing a user profile through User Management -> Click on user profile
	a. also pauses when clicking user dashboard (Red/Green Dot from leadership board)
        b. I guess whenever pages have to reload timer is paused?
2. Timer is not pausing after not touching computer for over 5 minutes, however it does pause when computer falls asleep/shut down.
	a. Notification does pop up that timer has been paused
3. Timer does run smoothly, I like the accessibility of manually editing timer countdown (it also remembers your previous input)
	a. Time limits set up so I can't go beyond 59 minutes, but I also can't go beyond 5 
        hours? 
	b. 0:0 issue (mentioned already)
	c. User is able to add time to timer in 15, 30, 1hr increments while actively running, no 
        manual entry permitted
4. When timer is up, timer is set back to 0 and time entry log pops up, notification bell
	a. User is able to add time to timer in 15, 30, 1hr increments while actively running, no 
        manual entry permitted unless resetting timer
5. Even when two pages of HGN server were up, timer is synced and running, paused correctly

Thank you for all your work on this new implementation!

**Comment by ypan710** [comment]
I've only seen the countdown timer but not the stopwatch. When I increase the timer from like 29:52 by 15 minutes, the new time is 45:00 minutes, not 44:52 minutes, before counting down starts. Vice versa when I decrease the timer from 29:52 by 15 minutes, the new time is 15:00 minutes, not 14:52 minutes.

Screen Shot 2023-04-06 at 4 31 14 PM

Screen Shot 2023-04-06 at 4 31 26 PM

**Comment by LucileTech** [comment]
To be able to run the front server I had to do “npm i --legacy-peer-deps” instead of “npm i” 

-The timer pauses when accessing a user profile through User Management -> Click on the user profile
- If I reload the page the Timer pauses
-Timer is not pausing after not touching the computer for over 5 minutes
-It does pause when the computer falls asleep/shut down and a notification does pop up that timer has been paused
-The timer goes until 9h45min, and the minimum possible to set up Is 15minutes
-Time increments by 15, 30, 1hr, No manual entry permitted 
-When two pages of the HGN server are up the timer pauses correctly
-When the timer is up notification bell, the time entry log pops up and the timer is set back to 0

However:

-It still has the minus issue mentioned by @AriaYu927 

Screenshot 2023-04-06 at 16 59 39

-When I click on minus to remove time to time left it goes to 15 again

-The timer does not pause when clicking the user dashboard (Red/Green Dot from the leadership board)

**Review by ToobaJamal — APPROVED** [review]
I have reviewed your PR; the timer works as intended and syncs on new tabs too. It does add the time correctly, if I add 15 minutes to 29:23 it is incremented to 44:23. No minus issue on my end as well.
The timer does not allow subtracting time when the time remaining on the stopwatch is equal to or less than 15 minutes.

https://user-images.githubusercontent.com/52610124/230560254-d0e9483b-110e-4779-b919-6d8b368f574a.mp4

**Review by vitorsaa2k — COMMENTED** [review]
The timer seems to work as intended. I've started the timer and started going trough all pages of the app and it continues to work, I've opened multiple tabs and the timer continued to work normally. When the time reach to 0, the modal appears just as showed in the image.

image

The only problem is that when its the first time the user will use the timer, the goal is set to 0, when it should be 15.

image

**Review by Lucas-E — APPROVED** [review]
Hey guys! I've just tested the PR and it looks awesome and works as exprected. I changed the tabs and it didn't stop the timer. When i refreshed the page, a popup showed, and the timer was paused. Great job!
Captura de tela de 2023-04-07 17-42-10
Captura de tela de 2023-04-07 17-40-18

**Comment by PrabhjotSembhi** [comment]
I am getting errors. Please resolve Package.json conflicts first. I think that I am facing this issue `Module not found: Can't resolve 'react-circular-progressbar'`
image

**Comment by harlleybastos** [comment]
> I am getting errors. Please resolve Package.json conflicts first. I think that I am facing this issue `Module not found: Can't resolve 'react-circular-progressbar'` image

Please, try to just delete the branch and pull it again using the step-by-step that I put on the description. If not works please let me know!

File: src/components/Header/Header.css
Comment (Vishvesh-Sheoran) [inline src/components/Header/Header.css:1]: I've made the following Improvements:

1. Removed duplicate width definition in _.header-wrapper_.
2. Fixed missing closing brace in _.owner-message_.
3. Simplified media query syntax by removing redundant _max-width_ keyword.
4. Replaced _position: relative_ with _position: absolute_ in _.redBackGroupHeader_.
5. Changed _display: flex_; _width: 0_; to _width: 0_; _margin-right: 0_; in _.timer-message-section_ for better readability.
6. Removed unnecessary width: 100%; in _.timer-message-section_.
7. Reordered CSS properties in alphabetical order for better maintainability.

Code:

.header-wrapper {
  height: 82px;
  width: clamp(100vw, 1vw + 0.1rem, 100%);
  margin-bottom: 1rem;
}

.navbar {
  z-index: 100;
  white-space: nowrap;
}

.timer-message-section {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 1rem;
}

.nav-links {
  align-items: center;
}

.redBackGroupHeader {
  z-index: 10;
  bottom: 0;
  right: 0;
  position: absolute;
  height: 30px;
  text-align: center;
  vertical-align: middle;
  background: #ff4d4f;
  border-radius: 40px;
  color: #fff;
  padding: 3px 6px;
  max-width: 39px;
  font-size: 1rem;
  min-width: 29px;
}

@media screen and (max-width: 1500px) {
  .dashboard-text-link {
    font-size: 14px;
  }
}

@media screen and (max-width: 1400px) {
  .timer-message-section {
    width: 0;
    margin-right: 0;
  }

  .owner-message {
    display: none;
  }
}

@media screen and (max-width: 1200px) {
  .header-wrapper {
    height: 56px;
  }
}

**Review by 20chen-7 — CHANGES_REQUESTED** [review]
Look great, and I like the timer end sound. one issue is that when I click the "+" or "-" button, the timer will stop. another one is in the popup dialogue, the clear and close button, the clear look good, but the close does not work.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/102696148/6bc17370-9576-44ae-9c19-e9d5e271ed83

**Review by nathanah — CHANGES_REQUESTED** [review]

**Review by 20chen-7 — CHANGES_REQUESTED** [review]
Everything looks good, the timer will increase /decrease by 15 minutes after clicking "+"/"-". This is one issue like  [nathanah](https://github.com/nathanah) said, the timer is inconsistent. Even it will be consistent soon. In the pop-up dialogue, the back button also has a similar issue.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/102696148/58d7e21a-fb8f-40b1-98fe-a447e9b889fc

**Comment by eduhorta** [comment]
The new timer looks great!
Here are my observations:

When starting the timer and then changing to a new tab, the stop button becomes unavailable but it still works when clicked. However, the stop button when clicking on the "watch icon" won't work.
After sometime both stop buttons become available on the new tab.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/de49c211-0d2e-4297-a495-dd5b32a12aa4

As mentioned by @nathanah, if you pause the timer and press "-15 min" resulting in a remaining time bellow 15 mins it resets to 15 mins and you lose the time spent.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/23545dd9-4383-4d37-8fa9-25ffabc016fd

If you open a new tab and close the current one before the new one loads, your Timer is paused.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/173a3645-e8d4-40dd-bf5f-e93458e7a2e9

When you click on the reset button, you get a modal to confirm the action. But, if you click on Stop -> "Clear and Close" it resets the timer without a confirmation. Wouldn't it be interesting to use the same modal component for confirmation to avoid a miss click?

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/aea7b5ec-9bbf-4377-8409-0ee0c9c1fb56

**Comment by rauleffting** [comment]
> 

Hi! Did you run the backend? You are the only one with this issue.

**Comment by rauleffting** [comment]
> Everything looks good, the timer will increase /decrease by 15 minutes after clicking "+"/"-". This is one issue like [nathanah](https://github.com/nathanah) said, the timer is inconsistent. Even it will be consistent soon. In the pop-up dialogue, the back button also has a similar issue.
> 
>  Screen.Recording.2023-05-31.at.9.42.21.AM.mov

Hi! Did you run the backend? You are the only one with this issue.

File: src/components/Timer/NewTimer.jsx [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+      const lastAccess = moment(message?.lastAccess);
+      const elapsedTime = moment.duration(now.diff(lastAccess)).asMilliseconds();
+      let remaining = message?.time - elapsedTime;
+
+      if (remaining <= 900000) {
+        alert('Timer cannot be set to less than fifteen minutes!');
+        return;
+      } else if (remaining <= 1800000) {
+        sendMessage(action.SET_GOAL.concat(time));
+      } else {
```
[the remark this answers is not in the document]
Comment (rauleffting) [reply src/components/Timer/NewTimer.jsx:156]: Great! It was fixed!

**Comment by rauleffting** [comment]
Hi, @eduhorta. The bug you found with the stop button was fixed. The rest is to behave like this. Jae's decision. Thank you!

**Review by OAkinbode — COMMENTED** [review]
Hello, 

The GUI looks beautiful. I found 3 issues that might need attention:

1. The stop button on the timer is temperamental. Sometimes it works and at other times it does not. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/ce109834-f0fd-43d4-a8fe-01aec3efcfb6

2. Closing the modal from the clear and close button resets the time to the initial value (I inputted a custom value for the timer) without warning. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/129c535f-deff-45f7-aaa4-ff22f5717732

3. With the custom time value, when I try to reduce the time by 15 minutes from a value of 22 minutes, it doesn't give me a warning that I can't do that. It just resets my time to 22 minutes. This may be a bug with the custom time inputs. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/475cecd2-082e-4f38-ad4c-8fc81cc5d512

**Review by stevenpaalz — COMMENTED** [review]
The timer appears to be working as intended.

One thought on improvement: When a user clicks pause on a running timer, the pause and stop buttons are replaced by the play button. This is a bit confusing in the case where a user may press pause and then want to stop before resuming again. It may be better to keep the stop button in place when the timer is paused.

 I also noticed that the stop button is grayed out but not disabled when the timer is running. I am unclear if we intend to have the button disabled but if so, then this should be disabled or if not then this should be not grayed out.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/121642083/2cce6b8c-cf68-473f-9ec9-dca07f4daa2a

**Comment by rauleffting** [comment]
**PLEASE, DO NOT SUGGEST NEW FEATURES AND CHANGES, JUST TEST AND VERIFY IF IT'S WORKING.**

**Comment by rauleffting** [comment]
Hi @OAkinbode. I think I've fixed all bugs you mentioned. Thank you very much!

**Comment by papiasharmin** [comment]
Hello @rauleffting, everything is working as intended, all the button are working perfectly. I am not sure about this things,
Is the hour option on modal fixed to 5 intentionally? and will the timer will always deduct 15 mins when started? mine is deducting 30mins though.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/86232710/40bd44ac-d14e-44cd-bb31-25c608f00da2

**Comment by rauleffting** [comment]
Hi @papiasharmin. This behavior is very odd. I'll take a look soon as possible! Thank you!

**Comment by OAkinbode** [comment]
> Hi @OAkinbode. I think I've fixed all bugs you mentioned. Thank you very much!

Hello @rauleffting. Well done on all the work. This is my update: 

I pulled the latest changes and restarted front and back, but it seems I am still encountering the time reset bug and the irresponsive close button. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/eb39beb8-ce24-4dc3-ac13-79e835dc4d5c

I think I might have stumbled on a new bug also. I had the timer on 34 minutes and could not reduce it by 15 minutes with the "-15 button". Reducing it by 15 minutes should leave us with 19 minutes on the timer which should be a valid time value, no?

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/6e6939b5-e90c-49e0-830d-ed99946094c2

I am checked in on both the back and front end with the latest pulls:
Screenshot 2023-06-01 at 1 50 49 PM

**Review by EvianTan — APPROVED** [review]

**Review by EvianTan — APPROVED** [review]

**Pull request merged**